x-pack/legacy/plugins/reporting/server/lib/get_user.ts

```diff
@@ -4,16 +4,19 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-export function getUserFactory(server) {
-  return async request => {
+import { Legacy } from 'kibana';
+import { ServerFacade } from '../../types';
+
+export function getUserFactory(server: ServerFacade) {
+  return async (request: Legacy.Request) => {
```

Comment:
This function takes a Legacy request because it is called from middleware. That means we get into this code before we get into any request handlers which is where we pick the properties for RequestFacade.

Reply:
@pgayvallet @restrry this means that I don't actually need to call `request.getRawRequest` - doing so made all the code break since this isn't an actual RequestFacade :)

Reply:
ddf8fac added that method back, but for a different usage

Reply:
Just thinking loudly: why not call this function from a route handler? NP routing doesn't provide a pre-middleware mechanism at the moment.

Reply:
> NP routing doesn't provide a pre-middleware mechanism at the moment.

Hm, if that is the case it will be a major blocker for Reporting migration.

@rudolf can you confirm whether we will be able to register routing middleware in the new platform?

Reply:
> @rudolf can you confirm whether we will be able to register routing middleware in the new platform?

ATM this is not something we are planing to provide. 

> Hm, if that is the case it will be a major blocker for Reporting migration.

Correct me if I'm wrong, but the whole pre-routing mechanism can easily be replaced by some wrapper around the actual route handlers. The usage would be really close to what you currently have and should not be that big of a change

Looking at your `getRouteConfigFactoryReportingPre` method and usages in `x-pack/legacy/plugins/reporting/server/routes/generate_from_jobparams.ts`, the pre-routing can be moved to an higher level. This would be something similar as what we are going in `src/core/server/http/router/error_wrapper.ts`

```typescript
  const getRouteConfig = () => {
    const getOriginalRouteConfig: GetRouteConfigFactoryFn = getRouteConfigFactoryReportingPre(
      server
    );
    const routeConfigFactory: RouteConfigFactory = getOriginalRouteConfig(
      ({ params: { exportType } }) => exportType
    );

    return {
      ...routeConfigFactory,
      validate: {
        params: Joi.object({
         [...]
      },
    };
  };

  // generate report
  server.route({
    path: `${BASE_GENERATE}/{exportType}`,
    method: 'POST',
    options: getRouteConfig(),
    handler: async (legacyRequest: Legacy.Request, h: ReportingResponseToolkit) => {
      [...]
      return response;
    },
  });
```

would become something like (also adapter to NP router format)

```typescript
  const getRouteConfig = () => {
    const getOriginalRouteConfig: GetRouteConfigFactoryFn = getRouteConfigFactoryReportingPre(
      server
    );
    const routeConfigFactory: RouteConfigFactory = getOriginalRouteConfig(
      ({ params: { exportType } }) => exportType
    );

    return {
      ...routeConfigFactory,
      validate: {
        params: schema.object({
         [...]
      },
    };
  };


const routeConfig = getRouteConfig()
router.post(
      {
        path: `${BASE_GENERATE}/{exportType}`,
        options: routeConfig.options
        validate: routeConfig.validate,
      },
      routeConfig.performPreAuth(async (ctx, req, res, preAuthResult) => {
        const request = makeRequestFacade(legacyRequest, preAuthResults);
        ....
      })
    );
```

The `performPreAuth` generated from `getRouteConfigFactoryReportingPre` would need some adaption from the existing code to execute all pre-handlers and either throw if one throws or return the map of resolved objects (basically what HAPI was doing itself before), but that doesn't sound like big changes, the actual pre-hook handlers would remains unchanged.

Please ping me if you want me to go a little more in-details with that.

Reply:
I'm ++ on any solution that will allow us to move further, and I don't see any problems with your proposals. I think there are a lot of ways to achieve the goal that the code is doing and pre-routing is just one of them.